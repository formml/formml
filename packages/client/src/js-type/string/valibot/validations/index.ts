import bool from './bool.js'
import datetime from './datetime.js'
import decimal from './decimal.js'
import num from './num.js'

export { bool, datetime, decimal, num }
