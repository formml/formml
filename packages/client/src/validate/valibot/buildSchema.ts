import { Field } from '@formml/dsl'

export default function buildSchema(formmlSchema: Field) {
  return formmlSchema
}
