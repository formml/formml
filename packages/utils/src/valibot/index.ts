export * from './schemas/required.js'
export * from './validations/isoDateTime.js'
export * from './validations/numerical.js'
export * from './validations/allowBlank.js'
export * from './validations/numeric-validations.js'
