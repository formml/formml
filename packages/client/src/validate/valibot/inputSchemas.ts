import * as v from 'valibot'

export const number = () => v.special<string>((input) => !isNaN(Number(input)))
