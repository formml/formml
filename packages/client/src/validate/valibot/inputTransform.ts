import { BigNumber } from 'bignumber.js'
import dayjs from 'dayjs'
import * as v from 'valibot'

export const toNum = (schema?: v.BaseSchema) =>
  v.transform<v.StringSchema, number>(
    v.string([v.custom((i) => !isNaN(Number(i)))]),
    Number,
    schema,
  )
export const toDatetime = (schema?: v.BaseSchema) =>
  v.transform<v.StringSchema, dayjs.Dayjs>(
    v.string([v.isoTimestamp()]),
    dayjs,
    schema,
  )
export const toBool = (schema?: v.BaseSchema) =>
  v.transform<v.StringSchema, boolean>(v.string(), Boolean, schema)
export const toDecimal = (schema?: v.BaseSchema) =>
  v.transform<v.StringSchema, BigNumber>(
    v.string([v.custom((i) => !new BigNumber(i).isNaN())]),
    (i) => new BigNumber(i),
    schema,
  )
