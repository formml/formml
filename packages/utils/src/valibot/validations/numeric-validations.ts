import { BigNumber } from 'bignumber.js'
import * as v from 'valibot'

// Known Limitation: The wrapped function's parameters are always not optional even if the original function accepts optional parameters.
// This is a limitation of the TypeScript's Parameters utility type.
function toNumeric<
  TValidator extends (...args: unknown[]) => v.BaseValidation<
    number,
    number,
    v.BaseIssue<number>
  > & {
    requirement: unknown
  },
>(
  validator: TValidator,
  requirement: (
    value: BigNumber | number,
    original: ReturnType<TValidator>['requirement'],
    args: Parameters<TValidator>,
  ) => boolean,
  run?: (
    this: ReturnType<TValidator> & {
      requirement: (value: BigNumber | number) => boolean
    },
    ...args: Parameters<ReturnType<TValidator>['_run']>
  ) => ReturnType<ReturnType<TValidator>['_run']>,
) {
  return <TInput extends BigNumber | number>(
    ...args: Parameters<TValidator>
  ): v.BaseValidation<TInput, TInput, v.BaseIssue<TInput>> => {
    const action = validator(...args)
    return {
      ...action,
      _run: run ?? action._run,
      requirement: (value: BigNumber | number) =>
        requirement(value, action.requirement, args),
    } as v.BaseValidation<TInput, TInput, v.BaseIssue<TInput>>
  }
}

export const finite = toNumeric(v.finite, (value, original) =>
  BigNumber.isBigNumber(value) ? value.isFinite() : original(value),
)
export const integer = toNumeric(v.integer, (value, original) =>
  BigNumber.isBigNumber(value) ? value.isInteger() : original(value),
)
export const multipleOf = toNumeric(
  v.multipleOf,
  (value, _, args) =>
    BigNumber.isBigNumber(value)
      ? value.mod(args[0]).eq(0)
      : value % args[0] === 0,
  function (dataset, config) {
    if (dataset.typed && !this.requirement(dataset.value)) {
      v._addIssue(this, 'multiple', dataset, config)
    }
    return dataset
  },
)
