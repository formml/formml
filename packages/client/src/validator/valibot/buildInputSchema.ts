import { Field } from '@formml/dsl'

import { assertNever } from '../../utils/assertNever.js'
import buildSchema from './buildSchema.js'
import * as t from './transform/index.js'

export default function buildInputSchema(formmlSchema: Field) {
  if (formmlSchema.type === 'text') {
    const typedSchema = buildSchema(formmlSchema)
    return typedSchema
  }

  if (formmlSchema.type === 'num') {
    const typedSchema = buildSchema(formmlSchema)
    return t.toNum(typedSchema)
  }

  if (formmlSchema.type === 'bool') {
    const typedSchema = buildSchema(formmlSchema)
    return t.toBool(typedSchema)
  }

  if (formmlSchema.type === 'datetime') {
    const typedSchema = buildSchema(formmlSchema)
    return t.toDatetime(typedSchema)
  }

  if (formmlSchema.type === 'decimal') {
    const typedSchema = buildSchema(formmlSchema)
    return t.toDecimal(typedSchema)
  }

  return assertNever`Unsupported field schema: ${formmlSchema}`
}
