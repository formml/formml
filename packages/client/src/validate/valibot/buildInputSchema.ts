import { Field } from '@formml/dsl'

import { assertNever } from '../../utils/assertNever.js'
import buildSchema from './buildSchema.js'
import * as i from './inputTransform.js'

export default function buildInputSchema(formmlSchema: Field) {
  if (formmlSchema.type === 'text') {
    return buildSchema(formmlSchema)
  }

  if (formmlSchema.type === 'num') {
    return i.toNum()
  }

  if (formmlSchema.type === 'bool') {
    return i.toBool()
  }

  if (formmlSchema.type === 'datetime') {
    return i.toDatetime()
  }

  if (formmlSchema.type === 'decimal') {
    return i.toDecimal()
  }

  return assertNever`Unsupported field type: ${formmlSchema.type}`
}
