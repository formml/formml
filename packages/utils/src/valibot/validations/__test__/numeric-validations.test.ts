import { BigNumber } from 'bignumber.js'
import * as v from 'valibot'

import { finite, integer, multipleOf } from '../numeric-validations.js'

describe('numeric-validations', () => {
  describe('finite', () => {
    describe('with number type', () => {
      test.each([
        123,
        0,
        -456,
        Number.MAX_SAFE_INTEGER,
        Number.MIN_SAFE_INTEGER,
      ])('should validate finite number %p', (input) => {
        // Arrange
        const schema = v.pipe(v.number(), finite(undefined))

        // Act
        const result = v.safeParse(schema, input)

        // Assert
        expect(result.success).toBe(true)
      })

      test.each([
        Infinity,
        -Infinity,
        Number.POSITIVE_INFINITY,
        Number.NEGATIVE_INFINITY,
      ])('should reject infinite number %p', (input) => {
        // Arrange
        const schema = v.pipe(v.number(), finite(undefined))

        // Act
        const result = v.safeParse(schema, input)

        // Assert
        expect(result.success).toBe(false)
        expect(result.issues).toMatchSnapshot()
      })
    })

    describe('with BigNumber type', () => {
      test.each([
        new BigNumber('123.456'),
        new BigNumber('0'),
        new BigNumber('-456.789'),
        new BigNumber('1e+308'),
        new BigNumber('-1e+308'),
      ])('should validate finite BigNumber %p', (value) => {
        // Arrange
        const schema = v.pipe(v.instance(BigNumber), finite(undefined))

        // Act
        const result = v.safeParse(schema, value)

        // Assert
        expect(result.success).toBe(true)
      })

      test.each([new BigNumber('Infinity'), new BigNumber('-Infinity')])(
        'should reject infinite BigNumber %p',
        (value) => {
          // Arrange
          const schema = v.pipe(v.instance(BigNumber), finite(undefined))

          // Act
          const result = v.safeParse(schema, value)

          // Assert
          expect(result.success).toBe(false)
          expect(result.issues).toMatchSnapshot()
        },
      )
    })
  })

  describe('integer', () => {
    describe('with number type', () => {
      test.each([
        123,
        0,
        -456,
        Number.MAX_SAFE_INTEGER,
        Number.MIN_SAFE_INTEGER,
      ])('should validate integer number %p', (input) => {
        // Arrange
        const schema = v.pipe(v.number(), integer(undefined))

        // Act
        const result = v.safeParse(schema, input)

        // Assert
        expect(result.success).toBe(true)
      })

      test.each([123.45, -456.78, 0.1, Math.PI])(
        'should reject non-integer number %p',
        (input) => {
          // Arrange
          const schema = v.pipe(v.number(), integer(undefined))

          // Act
          const result = v.safeParse(schema, input)

          // Assert
          expect(result.success).toBe(false)
          expect(result.issues).toMatchSnapshot()
        },
      )
    })

    describe('with BigNumber type', () => {
      test.each([
        new BigNumber('123'),
        new BigNumber('0'),
        new BigNumber('-456'),
        new BigNumber('1000000000000000'),
      ])('should validate integer BigNumber %p', (value) => {
        // Arrange
        const schema = v.pipe(v.instance(BigNumber), integer(undefined))

        // Act
        const result = v.safeParse(schema, value)

        // Assert
        expect(result.success).toBe(true)
      })

      test.each([
        new BigNumber('123.456'),
        new BigNumber('-456.789'),
        new BigNumber('0.1'),
        new BigNumber('3.14159'),
      ])('should reject non-integer BigNumber %p', (value) => {
        // Arrange
        const schema = v.pipe(v.instance(BigNumber), integer(undefined))

        // Act
        const result = v.safeParse(schema, value)

        // Assert
        expect(result.success).toBe(false)
        expect(result.issues).toMatchSnapshot()
      })
    })
  })

  describe('multipleOf', () => {
    describe('with number type', () => {
      test.each([
        { base: 5, value: 0 },
        { base: 5, value: 5 },
        { base: 5, value: 10 },
        { base: 5, value: -15 },
        { base: 3, value: 9 },
      ])(
        'should validate number multiple of $base: $value',
        ({ base, value }) => {
          // Arrange
          const schema = v.pipe(v.number(), multipleOf(base, undefined))

          // Act
          const result = v.safeParse(schema, value)

          // Assert
          expect(result.success).toBe(true)
        },
      )

      test.each([
        { base: 5, value: 2 },
        { base: 5, value: 7 },
        { base: 5, value: -13 },
        { base: 3, value: 10 },
      ])(
        'should reject number not multiple of $base: $value',
        ({ base, value }) => {
          // Arrange
          const schema = v.pipe(v.number(), multipleOf(base, undefined))

          // Act
          const result = v.safeParse(schema, value)

          // Assert
          expect(result.success).toBe(false)
          expect(result.issues).toMatchSnapshot()
        },
      )

      test.each([
        { base: 0.5, value: 1.0 },
        { base: 0.5, value: 1.5 },
        { base: 0.5, value: 2.0 },
      ])(
        'should validate decimal multiple of $base: $value',
        ({ base, value }) => {
          // Arrange
          const schema = v.pipe(v.number(), multipleOf(base, undefined))

          // Act
          const result = v.safeParse(schema, value)

          // Assert
          expect(result.success).toBe(true)
        },
      )

      test.each([
        { base: 0.1, value: 0.3 },
        { base: 0.1, value: 0.7 },
      ])(
        'should fail validation for decimal multiple due to precision issues: $base, $value',
        ({ base, value }) => {
          // Arrange
          const schema = v.pipe(v.number(), multipleOf(base, undefined))

          // Act
          const result = v.safeParse(schema, value)

          // Assert
          expect(result.success).toBe(false)
          expect(result.issues).toMatchSnapshot()
        },
      )
    })

    describe('with BigNumber type', () => {
      test.each([
        { base: 5, value: new BigNumber('25') },
        { base: 5, value: new BigNumber('0') },
        { base: 5, value: new BigNumber('-30') },
        { base: 3, value: new BigNumber('999') },
      ])(
        'should validate BigNumber multiple of $base: $value',
        ({ base, value }) => {
          // Arrange
          const schema = v.pipe(
            v.instance(BigNumber),
            multipleOf(base, undefined),
          )

          // Act
          const result = v.safeParse(schema, value)

          // Assert
          expect(result.success).toBe(true)
        },
      )

      test.each([
        { base: 5, value: new BigNumber('23') },
        { base: 5, value: new BigNumber('-17') },
        { base: 3, value: new BigNumber('100') },
      ])(
        'should reject BigNumber not multiple of $base: $value',
        ({ base, value }) => {
          // Arrange
          const schema = v.pipe(
            v.instance(BigNumber),
            multipleOf(base, undefined),
          )

          // Act
          const result = v.safeParse(schema, value)

          // Assert
          expect(result.success).toBe(false)
          expect(result.issues).toMatchSnapshot()
        },
      )

      test.each([
        { base: 0.5, value: new BigNumber('2.5') },
        { base: 0.5, value: new BigNumber('3.0') },
        { base: 0.1, value: new BigNumber('0.3') },
        { base: 0.1, value: new BigNumber('0.7') },
      ])(
        'should validate BigNumber decimal multiple of $base: $value',
        ({ base, value }) => {
          // Arrange
          const schema = v.pipe(
            v.instance(BigNumber),
            multipleOf(base, undefined),
          )

          // Act
          const result = v.safeParse(schema, value)

          // Assert
          expect(result.success).toBe(true)
        },
      )
    })
  })
})
