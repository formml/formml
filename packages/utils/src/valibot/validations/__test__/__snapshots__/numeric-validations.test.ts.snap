// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`numeric-validations > finite > with BigNumber type > should reject infinite BigNumber %p BigNumber{ s: -1, e: null, c: null } 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": null,
    "input": "-Infinity",
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid finite: Received BigNumber",
    "path": undefined,
    "received": "BigNumber",
    "requirement": [Function],
    "type": "finite",
  },
]
`;

exports[`numeric-validations > finite > with BigNumber type > should reject infinite BigNumber %p BigNumber{ s: 1, e: null, c: null } 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": null,
    "input": "Infinity",
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid finite: Received BigNumber",
    "path": undefined,
    "received": "BigNumber",
    "requirement": [Function],
    "type": "finite",
  },
]
`;

exports[`numeric-validations > finite > with number type > should reject infinite number %p -Infinity 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": null,
    "input": -Infinity,
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid finite: Received -Infinity",
    "path": undefined,
    "received": "-Infinity",
    "requirement": [Function],
    "type": "finite",
  },
]
`;

exports[`numeric-validations > finite > with number type > should reject infinite number %p -Infinity 2`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": null,
    "input": -Infinity,
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid finite: Received -Infinity",
    "path": undefined,
    "received": "-Infinity",
    "requirement": [Function],
    "type": "finite",
  },
]
`;

exports[`numeric-validations > finite > with number type > should reject infinite number %p Infinity 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": null,
    "input": Infinity,
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid finite: Received Infinity",
    "path": undefined,
    "received": "Infinity",
    "requirement": [Function],
    "type": "finite",
  },
]
`;

exports[`numeric-validations > finite > with number type > should reject infinite number %p Infinity 2`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": null,
    "input": Infinity,
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid finite: Received Infinity",
    "path": undefined,
    "received": "Infinity",
    "requirement": [Function],
    "type": "finite",
  },
]
`;

exports[`numeric-validations > integer > with BigNumber type > should reject non-integer BigNumber %p BigNumber{ s: -1, e: 2, c: [ 456, 78900000000000 ] } 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": null,
    "input": "-456.789",
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid integer: Received BigNumber",
    "path": undefined,
    "received": "BigNumber",
    "requirement": [Function],
    "type": "integer",
  },
]
`;

exports[`numeric-validations > integer > with BigNumber type > should reject non-integer BigNumber %p BigNumber{ s: 1, e: +0, c: [ 3, 14159000000000 ] } 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": null,
    "input": "3.14159",
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid integer: Received BigNumber",
    "path": undefined,
    "received": "BigNumber",
    "requirement": [Function],
    "type": "integer",
  },
]
`;

exports[`numeric-validations > integer > with BigNumber type > should reject non-integer BigNumber %p BigNumber{ s: 1, e: -1, c: [ 10000000000000 ] } 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": null,
    "input": "0.1",
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid integer: Received BigNumber",
    "path": undefined,
    "received": "BigNumber",
    "requirement": [Function],
    "type": "integer",
  },
]
`;

exports[`numeric-validations > integer > with BigNumber type > should reject non-integer BigNumber %p BigNumber{ s: 1, e: 2, c: [ 123, 45600000000000 ] } 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": null,
    "input": "123.456",
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid integer: Received BigNumber",
    "path": undefined,
    "received": "BigNumber",
    "requirement": [Function],
    "type": "integer",
  },
]
`;

exports[`numeric-validations > integer > with number type > should reject non-integer number %p -456.78 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": null,
    "input": -456.78,
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid integer: Received -456.78",
    "path": undefined,
    "received": "-456.78",
    "requirement": [Function],
    "type": "integer",
  },
]
`;

exports[`numeric-validations > integer > with number type > should reject non-integer number %p 0.1 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": null,
    "input": 0.1,
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid integer: Received 0.1",
    "path": undefined,
    "received": "0.1",
    "requirement": [Function],
    "type": "integer",
  },
]
`;

exports[`numeric-validations > integer > with number type > should reject non-integer number %p 3.141592653589793 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": null,
    "input": 3.141592653589793,
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid integer: Received 3.141592653589793",
    "path": undefined,
    "received": "3.141592653589793",
    "requirement": [Function],
    "type": "integer",
  },
]
`;

exports[`numeric-validations > integer > with number type > should reject non-integer number %p 123.45 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": null,
    "input": 123.45,
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid integer: Received 123.45",
    "path": undefined,
    "received": "123.45",
    "requirement": [Function],
    "type": "integer",
  },
]
`;

exports[`numeric-validations > multipleOf > with BigNumber type > should reject BigNumber not multiple of 3: BigNumber{ s: 1, e: 2, c: [ 100 ] } 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": "%3",
    "input": "100",
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid multiple: Expected %3 but received BigNumber",
    "path": undefined,
    "received": "BigNumber",
    "requirement": [Function],
    "type": "multiple_of",
  },
]
`;

exports[`numeric-validations > multipleOf > with BigNumber type > should reject BigNumber not multiple of 5: BigNumber{ s: -1, e: 1, c: [ 17 ] } 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": "%5",
    "input": "-17",
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid multiple: Expected %5 but received BigNumber",
    "path": undefined,
    "received": "BigNumber",
    "requirement": [Function],
    "type": "multiple_of",
  },
]
`;

exports[`numeric-validations > multipleOf > with BigNumber type > should reject BigNumber not multiple of 5: BigNumber{ s: 1, e: 1, c: [ 23 ] } 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": "%5",
    "input": "23",
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid multiple: Expected %5 but received BigNumber",
    "path": undefined,
    "received": "BigNumber",
    "requirement": [Function],
    "type": "multiple_of",
  },
]
`;

exports[`numeric-validations > multipleOf > with number type > should fail validation for decimal multiple due to precision issues: 0.1, 0.3 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": "%0.1",
    "input": 0.3,
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid multiple: Expected %0.1 but received 0.3",
    "path": undefined,
    "received": "0.3",
    "requirement": [Function],
    "type": "multiple_of",
  },
]
`;

exports[`numeric-validations > multipleOf > with number type > should fail validation for decimal multiple due to precision issues: 0.1, 0.7 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": "%0.1",
    "input": 0.7,
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid multiple: Expected %0.1 but received 0.7",
    "path": undefined,
    "received": "0.7",
    "requirement": [Function],
    "type": "multiple_of",
  },
]
`;

exports[`numeric-validations > multipleOf > with number type > should reject number not multiple of 3: 10 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": "%3",
    "input": 10,
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid multiple: Expected %3 but received 10",
    "path": undefined,
    "received": "10",
    "requirement": [Function],
    "type": "multiple_of",
  },
]
`;

exports[`numeric-validations > multipleOf > with number type > should reject number not multiple of 5: -13 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": "%5",
    "input": -13,
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid multiple: Expected %5 but received -13",
    "path": undefined,
    "received": "-13",
    "requirement": [Function],
    "type": "multiple_of",
  },
]
`;

exports[`numeric-validations > multipleOf > with number type > should reject number not multiple of 5: 2 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": "%5",
    "input": 2,
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid multiple: Expected %5 but received 2",
    "path": undefined,
    "received": "2",
    "requirement": [Function],
    "type": "multiple_of",
  },
]
`;

exports[`numeric-validations > multipleOf > with number type > should reject number not multiple of 5: 7 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": "%5",
    "input": 7,
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid multiple: Expected %5 but received 7",
    "path": undefined,
    "received": "7",
    "requirement": [Function],
    "type": "multiple_of",
  },
]
`;
